src/Dialogs/Admin/Report.vala

```diff
@@ -0,0 +1,376 @@
+			css_classes = {"selection-mode"}
+		};
+		rule_legal.toggled.connect (update_report);
+		rule_spam = new Gtk.CheckButton () {
+			group = rule_other,
+			active = report_category == API.Admin.Report.Category.SPAM,
+			css_classes = {"selection-mode"}
+		};
+		rule_spam.toggled.connect (update_report);
+		rule_violation = new Gtk.CheckButton () {
+			group = rule_other,
+			active = report_category == API.Admin.Report.Category.VIOLATION,
+			css_classes = {"selection-mode"}
+		};
+		rule_violation.toggled.connect (update_report);
+
+		rule_other_row = new Adw.ActionRow () {
+			title = API.Admin.Report.Category.OTHER.to_string (),
+			activatable_widget = rule_other,
+			sensitive = !report.action_taken
+		};
+		rule_other_row.add_prefix (rule_other);
+
+		rule_legal_row = new Adw.ActionRow () {
+			title = API.Admin.Report.Category.LEGAL.to_string (),
+			activatable_widget = rule_legal,
+			sensitive = !report.action_taken
+		};
+		rule_legal_row.add_prefix (rule_legal);
+
+		rule_spam_row = new Adw.ActionRow () {
+			title = API.Admin.Report.Category.SPAM.to_string (),
+			activatable_widget = rule_spam,
+			sensitive = !report.action_taken
+		};
+		rule_spam_row.add_prefix (rule_spam);
+
+		rule_violation_row = new Adw.ActionRow () {
+			title = API.Admin.Report.Category.VIOLATION.to_string (),
+			activatable_widget = rule_violation,
+			sensitive = !report.action_taken
+		};
+		rule_violation_row.add_prefix (rule_violation);
+
+		rule_group.add (rule_other_row);
+		rule_group.add (rule_legal_row);
+		rule_group.add (rule_spam_row);
+		rule_group.add (rule_violation_row);
+
+		rules_group = new Adw.PreferencesGroup () {
+			title = _("Violated Rules"),
+			visible = report_category == API.Admin.Report.Category.VIOLATION
+		};
+
+		if (accounts.active.instance_info.rules != null && accounts.active.instance_info.rules.size > 0) {
+			rules_buttons = new Gee.HashMap<string, Gtk.CheckButton> ();
+			string[] selected_rules_ids = {};
+			foreach (var rule in report.rules) {
+				selected_rules_ids += rule.id;
+			}
+
+			foreach (var rule in accounts.active.instance_info.rules) {
+				var checkbutton = new Gtk.CheckButton () {
+					css_classes = {"selection-mode"},
+					active = rule.id in selected_rules_ids
+				};
+				checkbutton.toggled.connect (update_report);
+				rules_buttons.set (rule.id, checkbutton);
+
+				var rule_row = new Adw.ActionRow () {
+					title = GLib.Markup.escape_text (rule.text).strip (),
+					activatable_widget = checkbutton,
+					use_markup = true,
+					sensitive = !report.action_taken
+				};
+				rule_row.add_prefix (checkbutton);
+				rules_group.add (rule_row);
+			}
+		}
+
+		var status_group = new Adw.PreferencesGroup () {
+			title = _("Reported Posts")
+		};
+
+		report.statuses.foreach (status => {
+			try {
+				status.formal.filtered = null;
+				status.formal.spoiler_text = null;
+				Widgets.Status widget = (Widgets.Status) status.to_widget ();
```

Comment:
In report UI's, do make sure to blur & greyscale media displayed to the user (ideally with toggles off for each independently), to avoid display harmful content directly to moderators

(mastodon's moderator tools currently only blur reusing the sensitive media mode, instead of also have greyscaling)

Looks like it took a heap of work to implement this; if there's APIs or things you'd need from the mastodon side for admin do let me know

Reply:
Thanks! I'll move that to another issue so I can keep track of it.

> if there's APIs or things you'd need from the mastodon side for admin do let me know

I'm not an instance admin (also why I didn't think of or notice the blur and greyscale features) so I really don't know what instance admins would need from a native client either. As can be seen by the edit on the pr body, I went in this with the thought that I would be implementing *all* admin and mod sections, only to discover that most are not accessible by the api (everything under administration, invites...). I'm not sure if that's something an admin would want to do outside of mastodon-web (and tbh I'd rather Tuba wasn't responsible for any of these :))

Again, I really don't know if these are something admins would want, but some of the stuff I remember not being able to do through the API:
- Memoriam 
- Change user email
- Change user password
- Audit logs
- Notes
- Add/remove posts from a report

---

I found a bug while implementing updating reports, where if there are rule violations with selected rules, you cannot change category. Mastodon-web will show it as if it was changed, but api-wise it hasn't:

[Screencast from 2024-06-24 05-54-33.webm](https://github.com/GeopJr/Tuba/assets/18014039/16e1738c-a816-4fbf-a00a-69665422b3bd)

I was running a some-versions-behind instance when testing so I don't know if it has been fixed on master. From Tuba's side, I work around that by removing the rules before changing the category

Reply:
This looks to maybe just be an awkward API without good validation for report category change, as that's the way Mastodon's own UI does it too: https://github.com/mastodon/mastodon/blob/4743657ba24e83c376e9f477fbf49114e6f09a57/app/javascript/mastodon/components/admin/ReportReasonSelector.jsx#L127

Reply:
I think these might be most likely to be able to be added to the API:

- Audit logs
- Notes (assuming Account Notes and Report Notes)
- Add/remove posts from a report

Reply:
Ah looks like the fix is still in master https://github.com/mastodon/mastodon/commit/9ce914cc897d2b1100cee8cb6d792a2283d877ec

Thanks for taking a look!

> I think these might be most likely to be able to be added to the API

Sounds good to me!

Reply:
>  I'm not sure if that's something an admin would want to do outside of mastodon-web (and tbh I'd rather Tuba wasn't responsible for any of these :))

I'd agree tbh, as it is, I'm surprised you've implemented so much. Building good administrative and moderator UIs can be challenging because the needs are often very different to those of regular users.